### PR TITLE
test: provide a subsystem to register the unittest module

### PR DESCRIPTION
**Pull request by keturn**
Many current ModuleTestingEnvironment errors (#4757) come from a failure to find a module associated with some stub Components included with engine-tests.

This subsystem registers the `unittest` module containing those classes with the ModuleManager before the engine is initialized.

It's used on the MTE side by adding it to the list of subsystems given to the builder: https://github.com/Terasology/ModuleTestingEnvironment/pull/55

## Pros

- Looks like we're on the right track. With this, MTE's own test suite goes from almost-all-failing to 19 pass, 3 failing.
- Very little code to implement and use.

## Cons

- 💄🐷. I still feel this is a kludge that isn't exactly making things simpler. At least it using an existing extension point (subsystems), but it adds a new thing to the mix that isn't usually considered in the module-loading process.
- I worry it's fragile. This makes things work because Subsystem.initialize is called at a useful place in the engine's boot sequence, but it's very possible that something else will call `ModuleManager.loadEnvironment` later—_without_ including `"unittest"` in the environment—and then things will break with an unclear error again. (At least, I think so. Might not. Should test.)
- It leaves the question of "why are we trying to register all components before loading modules?" unanswered.

## Outstanding but maybe not critical?
- Question: If we register the module _without_ entering it in the current module environment, is that sufficient to avoid the error?
- Are there other places in engine-test that should prefer this over their current use of ModuleManagerFactory?

### CODE REVIEW

File: engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+package org.terasology.engine.testUtil;
+
+import org.terasology.engine.context.Context;
+import org.terasology.engine.core.GameEngine;
+import org.terasology.engine.core.module.ModuleManager;
+import org.terasology.engine.core.subsystem.EngineSubsystem;
+import org.terasology.gestalt.module.Module;
+
+public class WithUnittestModule implements EngineSubsystem {
```
Comment (keturn) [inline engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java:12]: any suggestions for the name? should it end it `Subsystem`?

File: engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public String getName() {
+        return "Unittest";
+    }
+
+    @Override
+    public void initialise(GameEngine engine, Context rootContext) {
+        EngineSubsystem.super.initialise(engine, rootContext);
```
Comment (keturn) [inline engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java:20]: Calling `super` out of habit here, but the default implementation is a no-op. Is that a good practice or should we remove this `super`?

**Comment by keturn** [comment]
Switched to _draft_ until we get the remaining tests fixed— or find out they are failing for an unrelated reason.

**Comment by keturn** [comment]
All three test failures were in [TestEventReceiverTest](https://github.com/Terasology/ModuleTestingEnvironment/blob/2bccc60d48be1ce07628f3bfa4ce9046f27e3447/src/test/java/org/terasology/moduletestingenvironment/TestEventReceiverTest.java#L26) and they all fail in `onDummyEvent`:
```java
@Share(DummySystem.class)
@RegisterSystem(RegisterMode.AUTHORITY)
public class DummySystem extends BaseComponentSystem {
    @ReceiveEvent
    public void onDummyEvent(DummyEvent event, EntityRef entity) {
        DummyComponent component = entity.getComponent(DummyComponent.class);
        component.dummy = true;
        entity.saveComponent(component);
    }
}
```
with a NullPointerException suggesting that the `entity.getComponent` call returned null.

**Comment by keturn** [comment]
I am [confused](https://discord.com/channels/270264625419911192/696830442065625118/858128587332255785) about those test failures, but I'm willing to say they're not necessarily within the scope of this PR.

**Comment by keturn** [comment]
so far I have only checked MTE's own test suite.

Still need to check the MTE tests of other modules.

and once this is merged (with MTE#55), then we'll start seeing if that was sufficient for the errors when run under the CI environment, or if that's something else.

**Comment by keturn** [comment]
I ran some module tests. There are some failures, I haven't yet tallied them to see what the failure modes are: https://scans.gradle.com/s/6vj4ag42fug5w/tests/failed

**Comment by keturn** [comment]
Four modules:
- Pathfinding
- NameGenerator
- SimpleFarming
- Tasks

so it's worth noting that a lot of other modules with MTE tests (like Health's RegenTest) passed!

Tasks is the smallest of those four, so that's the thing I'm looking at next.

**Comment by keturn** [comment]
`Tasks` was pulling in an old version of MTE. I fixed that.

The others are using various non-MTE tests and so we can call them out of scope for this PR.

File: engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java (hunk 1 above)
Comment (skaldarnar) [reply engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java:12]: A survey of what we currently (including this PR) have shows:

- Usage in extends/implements clause: 22
- 18 with `Subsystem` suffix / 5 without suffix
- classes without suffix are
	- WithUnittestModule
	- OpenVRInput 
	- Headless[Audio|Graphics|Input]
	
Given that, I'd suggest to
- add the `Subsystem` suffix here. Would `RegisterUnittestModuleSubsystem` be a reasonable name?
- (different PR) rename `OpenVRInput` >>> `OpenVrInputSubsystem`
- (different PR) add `Subsystem` suffix to `Headless[Audio|Graphics|Input]`
- (different PR) make sure to consistently use `Subsystem` (we use `SubSystem` in some places)

The latter changes would fit the topic of code style improvements @jdrueckert :wink:

File: engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java (hunk 2 above)
Comment (skaldarnar) [reply engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java:20]: I think I would not make a call to `super` when overriding _default_ methods of an interface usually :thinking:
I understand the `EngineSubsystem` interface as something where I get a bunch of hook methods I can choose to implement, but I don't have to. 

If it suddenly becomes necessary to always perform some kind of code here you'd probably change the underlying implementation such that contract stays the same for child classes (e.g., introduce a new "internal init method" that in turn calls the hook method...

tl;dr: I don't think it is necessary, and I would omit it.

File: engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java (hunk 1 above)
Comment (skaldarnar) [reply engine-tests/src/main/java/org/terasology/engine/testUtil/WithUnittestModule.java:12]: I'd appreciate a short docstring comment for this subsystem explaining why we have it, and what it is used for.

**Review by skaldarnar — APPROVED** [review]

**Review by skaldarnar — APPROVED** [review]
